core/pkg/server/config.go

```diff
@@ -0,0 +1,236 @@
+// used to store programmatic custom charts for the run.
+//
+// The server process builds this up incrementally throughout a run's lifetime.
+type RunConfig struct {
+	// The underlying configuration tree.
+	//
+	// Nodes are strings and leaves are types supported by JSON,
+	// such as primitives and lists.
+	tree RunConfigDict
+}
+
+type ConfigFormat int
+
+const (
+	FORMAT_YAML ConfigFormat = iota
+	FORMAT_JSON
+)
+
+func NewRunConfig() *RunConfig {
+	return &RunConfig{make(RunConfigDict)}
+}
+
+func NewRunConfigFrom(tree RunConfigDict) *RunConfig {
+	return &RunConfig{tree}
+}
+
+// Returns the underlying config tree.
+//
+// Provided temporarily as part of a refactor. Avoid using this, especially
+// mutating it.
+func (runConfig *RunConfig) Tree() RunConfigDict {
+	return runConfig.tree
+}
+
+// Updates and/or removes values from the configuration tree.
+//
+// Does a best-effort job to apply all changes. Errors are passed to `onError`
+// and skipped.
+func (runConfig *RunConfig) ApplyChangeRecord(
+	configRecord *service.ConfigRecord,
+	onError func(error),
+) {
+	for _, configItem := range configRecord.Update {
+		path := keyPath(configItem)
+
+		var value interface{}
+		if err := json.Unmarshal(
+			[]byte(configItem.GetValueJson()),
+			&value,
+		); err != nil {
+			onError(
+				fmt.Errorf(
+					"Failed to unmarshall JSON for config key %v: %w",
+					path,
+					err,
+				),
+			)
+			continue
+		}
+
+		if err := runConfig.updateAtPath(path, value); err != nil {
+			onError(err)
+			continue
+		}
+	}
+
+	for _, configItem := range configRecord.Remove {
+		runConfig.removeAtPath(keyPath(configItem))
+	}
+}
+
+// Inserts W&B-internal values into the run's configuration.
+func (runConfig *RunConfig) AddTelemetryAndMetrics(
+	telemetry *service.TelemetryRecord,
+	metrics []map[int]interface{},
+) {
+	wandbInternal := runConfig.internalSubtree()
+
+	if telemetry.GetCliVersion() != "" {
+		wandbInternal["cli_version"] = telemetry.CliVersion
+	}
+	if telemetry.GetPythonVersion() != "" {
+		wandbInternal["python_version"] = telemetry.PythonVersion
+	}
+
+	wandbInternal["t"] = corelib.ProtoEncodeToDict(telemetry)
+
+	if metrics != nil {
+		wandbInternal["m"] = metrics
+	}
+}
+
+// Serializes the run configuration to send to the backend.
+func (runConfig *RunConfig) Serialize(format ConfigFormat) ([]byte, error) {
+	// A configuration dict in the format expected by the backend.
+	valueConfig := make(map[string]map[string]interface{})
+	for treeKey, treeValue := range runConfig.tree {
+		valueConfig[treeKey] = map[string]interface{}{
+			"value": treeValue,
+		}
+	}
+
+	switch format {
+	case FORMAT_YAML:
+		return yaml.Marshal(valueConfig)
+	case FORMAT_JSON:
+		return json.Marshal(valueConfig)
+	}
+
+	return nil, fmt.Errorf("Unknown format: %v", format)
+}
+
+func (runConfig *RunConfig) internalSubtree() RunConfigDict {
+	node, found := runConfig.tree["_wandb"]
+
+	if !found {
```

Comment:
should we test this code path?

Reply:
good callout, added

ideally it'd be covered by resume tests, but i added a test for now (it's an OK test, just not the most valuable)